llvm/lib/Target/DirectX/DXILDataScalarization.cpp

```diff
@@ -305,6 +305,33 @@ bool DataScalarizerVisitor::visitGetElementPtrInst(GetElementPtrInst &GEPI) {
   Type *OrigGEPType = GEPI.getSourceElementType();
   Type *NewGEPType = OrigGEPType;
   bool NeedsTransform = false;
+  // Check if the pointer operand is a ConstantExpr GEP
+  if (auto *PtrOpGEPCE = dyn_cast<ConstantExpr>(PtrOperand);
+      PtrOpGEPCE && PtrOpGEPCE->getOpcode() == Instruction::GetElementPtr) {
+    if (GlobalVariable *NewGlobal =
```

Comment:
Note: The pointer operand of a ConstExpr GEP can be another ConstExpr GEP.

Example: https://github.com/llvm/llvm-project/blob/66e707ec68d6dc8d8f70083b68daf52559a862c1/llvm/test/CodeGen/DirectX/flatten-array.ll#L136-L142

Reply:
I don't think it will ever naturally occur though. It may suffice to assert that the ConstantExpr GEP's pointer operand is a global variable and deal with it later if it does become an actual problem.

AFAIK there isn't a good reason to codegen a multiply-nested ConstantExpr GEP in the first place when you can codegen a single ConstantExpr GEP instead because all the indices are ConstantInts.